Source/WebCore/platform/text/WritingMode.h

```diff
@@ -31,233 +31,397 @@
+    constexpr inline bool isInlineLeftToRight() const;
+    // Block OR inline flow is top-to-bottom.
+    constexpr inline bool isAnyTopToBottom() const;
+    // Block OR inline flow is left-to-right.
+    constexpr inline bool isAnyLeftToRight() const;
+
+    // Typesetting modes.
+    constexpr inline bool isVerticalTypographic() const;
+    constexpr inline bool prefersCentralBaseline() const;
+    constexpr inline bool isMixedOrientation() const;
+    constexpr inline bool isUprightOrientation() const;
+    constexpr inline bool isSidewaysOrientation() const;
+
+    // Line orientation.
+    constexpr inline bool isLineInverted() const;  // line-over != block-start
+    constexpr inline bool isLineOverRight() const; // line-over == right
+    constexpr inline bool isLineOverLeft() const;  // line-over == left
+
+    // Coordinate flow queries.
+    constexpr inline bool isLogicalLeftLineLeft() const; // Not true for sideways-lr.
+    constexpr inline bool isLogicalLeftInlineStart() const; // == !isInlineFlipped()
+    constexpr inline bool isInlineFlipped() const; // Inline direction is RTL or BTT.
+    constexpr inline bool isBlockFlipped() const;  // Block direction is RL or BT.
+
+    // Directions as enums. Prefer booleans if doing boolean checks.
+    constexpr inline FlowDirection blockDirection() const;
+    constexpr inline TextDirection bidiDirection() const;
+    constexpr inline FlowDirection inlineDirection() const;
+
+    // Computed values. May differ from used values above.
+    constexpr inline WritingModeStyle computedWritingMode() const;
+    constexpr inline TextDirection computedTextDirection() const;
+    constexpr inline TextOrientation computedTextOrientation() const;
+    // ^ USE THESE FOR STYLE COMPUTATION ^ //
+
+    // Setters.
+    inline void setWritingMode(WritingModeStyle);
+    inline void setTextOrientation(TextOrientation);
+    inline void setTextDirection(TextDirection);
+
+    // To aid in packing.
+    using Data = uint8_t;
+    Data toData() const { return m_bits; }
+    constexpr inline WritingModeData(Data bits = 0) { m_bits = bits; }
+
+    friend bool operator==(WritingModeData, WritingModeData) = default;
+
+private:
+    Data m_bits { 0 };
+
+public: // Private except WritingModeStyle and FlowDirection are friends
+    enum Bits {
+        kIsVerticalText  = 1 << 0, // Vertical writing modes.
+        kIsFlippedBlock  = 1 << 1, // RL or BT block flow directions.
+        kIsVerticalType  = 1 << 2, // Vertical typographic mode.
+        kIsBidiRTL       = 1 << 3, // Bidi directionality.
+        kIsUprightType   = 1 << 4, // Upright text orientation.
+        kIsSidewaysType  = 1 << 5, // Sideways text orientation.
+
+        kBlockFlowMask   = kIsVerticalText | kIsFlippedBlock,
+        kWritingModeMask = kBlockFlowMask  | kIsVerticalType,
+        kOrientationMask = kIsSidewaysType | kIsUprightType, // Both is an error.
+        kOrientationShift = 4,
+    };
```

Comment:
Is that actually going to be more readable given all the bitwise operations like

```
    return (m_bits & (kIsVerticalText | kIsFlippedBlock)) == kIsVerticalText;
```

and

```
    return (m_bits & (kIsVerticalText | kIsFlippedBlock)) == kIsVerticalText;
```

though? All of these constants would need to be constructed as OptionSet objects, right? It seems like a lot of noise to me.

Reply:
(And the alternative of casting out m_bits in order to do bit math on it is... well that's literally all we're doing, so what's the point of storing it as an OptionSet in that case.)

Reply:
Keep in mind this is an entirely internal data representation. Nobody using the WritingMode object should ever need to know whether it's backed by bits or OptionSet or whatever. So the goal should be readability of these internal functions; the external world doesn't care.

Reply:
Right. We usually opt for readability for internal presentations too, until there is evidence that doing something else is actually beneficial (and OptionSet by itself is zero cost anyway). But I suppose bit fiddling may make sense here since these functions are invoked a lot.